Bottom: 42f51fa788dbddca2bd9fec1528daedc98ff093a
Top:    a883fc34b9c2c40f9824baf42576ddcb00500ed7
Author: scott snyder <sss@karma>
Date:   2020-05-30 00:32:22 -0400

Fix for Ctrl-Alt-Del (handle Starcaster Chronicles).



---

diff --git a/ComicReader/src/com/blogspot/applications4android/comicreader/comics/CtrlAltDel.java b/ComicReader/src/com/blogspot/applications4android/comicreader/comics/CtrlAltDel.java
index 524b305..bd3f185 100755
--- a/ComicReader/src/com/blogspot/applications4android/comicreader/comics/CtrlAltDel.java
+++ b/ComicReader/src/com/blogspot/applications4android/comicreader/comics/CtrlAltDel.java
@@ -88,6 +88,18 @@ public class CtrlAltDel extends DailyComic {
 
   
         protected Calendar getTimeFromUrl(String url)
+        {
+          Calendar c = getTimeFromUrl (url, 0);
+          if (c != null) {
+            return c;
+          }
+          CADStrip s = getCADStripFromUrl (url);
+          if (s != null) {
+            return s.mCal;
+          }
+          return null;
+        }
+        protected Calendar getTimeFromUrl(String url, int next)
         {
           int i = url.lastIndexOf ("ENG_");
           if (i >= 0) {
@@ -107,8 +119,32 @@ public class CtrlAltDel extends DailyComic {
                 i = url.lastIndexOf ("Strip");
                 if (i >= 0) {
                   i += 5;
+                  int j = url.lastIndexOf ("_");
+                  if (j >= 0 && j > i) {
+                    //https://cad-comic.com/wp-content/uploads/2020/05/StripXX88_2020515a.x44380.png
+                    i = j+1;
+                    if (url.substring(i+7,i+8).equals ("a")) {
+                      int y = Integer.parseInt (url.substring (i,   i+4));
+                      int m = Integer.parseInt (url.substring (i+4, i+5));
+                      int d = Integer.parseInt (url.substring (i+5, i+7));
+                      Calendar c = Calendar.getInstance();
+                      c.set (y, m-1, d);
+                      return c;
+                    }
+                  }
                 }
                 else {
+                  //https://cad-comic.com/wp-content/uploads/2020/04/SC07_17.x44380.png
+                  i = url.lastIndexOf ("/SC");
+                  if (i >= 0) {
+                    if (next == 0) return null;
+                    int y = Integer.parseInt (url.substring (i-7,   i-3));
+                    int m = Integer.parseInt (url.substring (i-2,   i));
+                    int d = next;
+                    Calendar c = Calendar.getInstance();
+                    c.set (y, m-1, d);
+                    return c;
+                  }
                   return null;
                 }
               }
@@ -123,37 +159,6 @@ public class CtrlAltDel extends DailyComic {
         }
 
 
-        private Calendar guessTimeFromUrl(String url)
-        {
-           // https://..../ENG_201811121.png
-           int i = url.lastIndexOf ("ENG_");
-           if (i >= 0) {
-             i += 4;
-           }
-           else {
-             i = url.lastIndexOf ("sillies-");
-             if (i >= 0) {
-               i += 8;
-             }
-             else {
-               i = url.lastIndexOf ("cad-");
-               if (i >= 0) {
-                 i += 4;
-               }
-               else {
-                 return null;
-               }
-             }
-           }
-           int y = Integer.parseInt (url.substring (i,   i+4));
-           int m = Integer.parseInt (url.substring (i+4, i+6));
-           int d = Integer.parseInt (url.substring (i+6, i+8));
-           Calendar c = Calendar.getInstance();
-           c.set (y, m-1, d);
-           return c;
-        }
-
-
         private String monthKey(Calendar cal)
         {
           int y = cal.get (Calendar.YEAR);
@@ -263,8 +268,20 @@ public class CtrlAltDel extends DailyComic {
 
         protected CADStrip getCADStripFromUrl(String url)
         {
-          Calendar c = getTimeFromUrl (url);
-          return getStripFromTime (c);
+          Calendar c = getTimeFromUrl (url, 0);
+          if (c != null) {
+            return getStripFromTime (c);
+          }
+          //https://cad-comic.com/wp-content/uploads/2020/04/SC07_17.x44380.png
+          int i = url.lastIndexOf ("/SC");
+          if (i >= 0) {
+            String yyyymm = url.substring (i-7,   i-3) + url.substring (i-2,   i);
+            SortedMap<Integer, CADStrip> map = getUrlsForMonth (yyyymm);
+            for (CADStrip s : map.values()) {
+              if (s.mUrl.equals (url)) return s;
+            }
+          }
+          return null;
         }
 
 
@@ -320,15 +337,17 @@ public class CtrlAltDel extends DailyComic {
                   JSONObject root = new JSONObject(sdata);
                   JSONArray posts = root.getJSONArray("posts");
                   List<String> ret = new ArrayList<String>();
-                  for (int i = 0; i < posts.length(); i++) {
+                  int next = 1;
+                  for (int i = posts.length()-1; i >= 0; i--) {
                     JSONObject o = posts.getJSONObject(i);
                     String url = o.getString("comic");
                     if (url.lastIndexOf("KDM") >= 0) continue;
-                    Calendar c = getTimeFromUrl (url);
+                    Calendar c = getTimeFromUrl (url, next);
                     if (!month.equals (monthKey (c))) {
                       continue;
                     }
                     int dom = c.get (Calendar.DAY_OF_MONTH);
+                    next = dom+1;
                     CADStrip strip = new CADStrip (url, o.getString("title"), o.getString("date"), c);
                     l.put (dom, strip);
                   }
